### ISSUE TITLE
import prefer `js` over `css/sass/scss`

### ISSUE THREAD

**Issue by larpo** [webpack-contrib/sass-loader#556]
It appears that 7.0 breaks sass modules we are using from npm. I guess it is related to https://github.com/webpack-contrib/sass-loader/pull/479

For example foundation-sites that we are using imports it's internal modules like this:
```
// Sass utilities
@import 'util/util';
```

Now, unfortunately 'util/util' is a valid file inside node_modules (from 'node-util' pkg), and the build fails with errror like this:

```
ERROR in ./src/style.scss (./node_modules/css-loader!./node_modules/sass-loader/lib/loader.js!./src/style.scss)
Module build failed: 
// Copyright Joyent, Inc. and other Node contributors.
^
      Invalid CSS after "...N THE SOFTWARE.": expected 1 selector or at-rule, was "var formatRegExp = "
      in ###/webpack-sass-broken-imports/node_modules/util/util.js (line 1, column 1)
 @ ./src/style.scss 2:14-117
 @ ./src/index.js
```
repro repo: https://github.com/larpo/webpack-sass-broken-module-imports

Is there a way to revert to the old behaviour? I don't see any mentions on READMe about how to configure the aliasing behaviour.

**Comment by evilebottnawi** [comment]
@larpo Confirm, wip

**Comment by jhnns** [comment]
Yep, fix is one the way. Thanks for the test case 👍

**Comment by evilebottnawi** [comment]
@jhnns can you take care?

**Comment by larpo** [comment]
Good to know it is a regression and not the intended behaviour :)

**Comment by jhnns** [comment]
Shipped as `7.0.1` 🚀.

That's why reproducible test cases are so good 👍

**Comment by evilebottnawi** [comment]
@jhnns looks like we still have problem :confused:

**Comment by evilebottnawi** [comment]
@jhnns found problem:
I have in directory two file with same name `component.js` and `component.scss` , also have `app.scss` with code
```
@import "component";
```
And `const request = utils.urlToRequest(url);` return file with js

**Comment by evilebottnawi** [comment]
@jhnns looks like problem here https://github.com/webpack-contrib/sass-loader/blob/master/lib/loader.js#L61, if you have two file with difference extensions, resolve prefer `.js` :confused:

**Comment by julmot** [comment]
Hello guys,
I'm coming here since I'm also experiencing issues in this PR about updating to v7:
https://github.com/julmot/form-components/pull/14#issuecomment-381153653
The build fails now with an error that looks like an issue of the loader to me: https://travis-ci.org/julmot/form-components/builds/366135805#L4146

**Comment by dak** [comment]
Just confirming I'm seeing the same issue as @evilebottnawi

**Comment by evilebottnawi** [comment]
I have PR https://github.com/webpack-contrib/sass-loader/pull/558, Can you test this and confirm what is fixed your problem

**Comment by julmot** [comment]
@evilebottnawi Yes, the build passes with this branch.

**Comment by dpikt** [comment]
@evilebottnawi PR #558 fixes the issue for me as well.

**Comment by ioanungurean** [comment]
@larpo the loader works now for me too. I had errors with `v7.0.0` and `v7.0.1`. 

Use: `npm install git://github.com/webpack-contrib/sass-loader.git#fix-resolve-conflict --save-dev` as @julmot pointed out. 

Thank you for the PR @evilebottnawi!

**Comment by julmot** [comment]
Please let us know when the PR is merged.

**Comment by ioanungurean** [comment]
What it's taking so long to be reviewed? The PR was open 10 day ago. :cry:

**Comment by phahn** [comment]
This is also causing issues for me in facebook/create-react-app#3815

**Comment by jhnns** [comment]
We need to go a different route (see [here](https://github.com/webpack-contrib/sass-loader/pull/558#issuecomment-384263644)). I'm thinking about making webpack's resolver opt-in and moving back to Sass' regular behavior.

Until then you should use webpack's [`resolve.modules`](https://webpack.js.org/configuration/resolve/#resolve-modules) option instead of node-sass' `includePaths` option if you experience any problems.

**Comment by julmot** [comment]
Ping guys

**Comment by ioanungurean** [comment]
I did not upgrade to `^7.0.0` because my build breaks. I'm staying with `6.0.7` for now.

**Comment by julmot** [comment]
I do too, but I'd rather like to have a solution for this issue (not a workaround) instead of keeping the old version.

**Comment by aclion111** [comment]
I'm setting up a new site, and bumped into this issue again.

Are folks here still sticking to v6.0.7 to avoid breakage?

**Comment by jsg2021** [comment]
yes, at least i am.

**Comment by ioanungurean** [comment]
Yes, unfortunately. :disappointed:

**Comment by evilebottnawi** [comment]
WIP on this, but it is really hard to solve keeping aliases

**Comment by evilebottnawi** [comment]
Workaround: using always `.sass/.scss` extension on any import inside `sass/scss` files

**Comment by emilio-martinez** [comment]
Bumped into this issue. In my case, I was using the node-sass `includePaths` option rather than the webpack resolution (we have other use cases that don't involve building with webpack). I was able to resolve by including the corresponding extension when referencing the problematic `.sass/.scss` files.

**Comment by evilebottnawi** [comment]
FMI:
hack how we can solve problem, but with limitation, aliasing for `css/scss/sass` only support alias with extension `css/sass/scss`, but we can add this to readme
```
sokra [12 minutes ago]
`this._compilation.resolverFactory.get("normal", { ...this._module.resolveOptions, extensions: [...] })`
```

**Comment by jsg2021** [comment]
@jhnns I was under the impression the normal sass/css imports happened by default, and node_module's needed the import with `~` prefix to use webpack's resolver... is this not the case anymore?

**Comment by evilebottnawi** [comment]
@jhnns we can fix this using advice from @sokra :
```
this._compilation.resolverFactory.get("normal", { ...this._module.resolveOptions, extensions: [...] })
```
And write in docs what aliasing working only with `css/sass/scss`. We have a lot of project where use `alias` and using `includePaths` to allow working some legacy sass libraries and frameworks

**Comment by emilio-martinez** [comment]
Perhaps something relevant to this conversation if Webpack is not doing the module resolution:

As of [`node-sass@4.9.0`](https://github.com/sass/node-sass/releases/tag/v4.9.0)

> `@import`ing of `.css` files will produce a deprecation warning. This behaviour will be removed in node-sass@5.0.0.
> 
> This behaviour violates the Sass language specification. Additionally it makes it impossible to output to the input directory via the CLI or watcher.

**Comment by evilebottnawi** [comment]
@larpo because it is works only for modules, i.e. `~module/path/to/file.scss`

**Comment by julmot** [comment]
So, what would be the workaround then?

**Comment by evilebottnawi** [comment]
@julmot https://github.com/webpack-contrib/sass-loader/issues/556#issuecomment-394019055

**Comment by julmot** [comment]
@evilebottnawi Okay, thanks. If this is the only workaround – modify the source code (import) in all imports – for this bug I think I'm going to prefer to keep the old version. Please keep us posted when this is resolved without modifying the source.

**Comment by evilebottnawi** [comment]
@julmot it is in my todo, i think it will be solve in this week :+1:

**Comment by evilebottnawi** [comment]
@julmot in your example you forget `resolve.extensions: [".scss", ".sass", ".css", "list-of-your-extensions-or-standard-js-and-json"]`

**Comment by julmot** [comment]
I still can't upgrade to sass-loader v7 due to this bug.

**Comment by evilebottnawi** [comment]
@julmot can you create minimum reproducible test repo?

**Comment by joshdcomp** [comment]
In case it can help someone else, this worked for me:

In your Webpack config's `module.rules[]`, move your rules for non-js assets to be evaluated before the js, so:

```js
module: {
    rules: [
      {
        test: /\.js$/,
        exclude: /node_modules/,
        use: {
          loader: 'babel-loader'
        }
      },
      {
        test: /\.(ttf|eot|woff|woff2|svg)$/,
        use: {
          loader: "file-loader",
          options: {
            name: "[name].[ext]",
          },
        },
      },
      {
        test: /\.scss$/,
        use: ExtractTextPlugin.extract({
          fallback: "style-loader",
          use: ['css-loader', 'sass-loader']
        })
      }
    ]
  },
```
becomes:

```js
module: {
    rules: [
      {
        test: /\.(ttf|eot|woff|woff2|svg)$/,
        use: {
          loader: "file-loader",
          options: {
            name: "[name].[ext]",
          },
        },
      },
      {
        test: /\.scss$/,
        use: ExtractTextPlugin.extract({
          fallback: "style-loader",
          use: ['css-loader', 'sass-loader']
        })
      },
      {
        test: /\.js$/,
        exclude: /node_modules/,
        use: {
          loader: 'babel-loader'
        }
      },
    ], 
  },
```

**Comment by nataliawojtkowska** [comment]
+1

**Comment by antonellil** [comment]
experiencing the same JS precendence issue as well, currently unable to upgrade to `7.*.*` from `6.0.7`

**Comment by larpo** [comment]
Not sure what you guys are doing differently, but this has been working perfectly for me since the fix in 7.0.1:

```
const srcPath = path.resolve(__dirname, "src/");
const aliases = fs
    .readdirSync(srcPath)
    .filter(file => fs.statSync(path.resolve(srcPath, file)).isDirectory())
    .reduce((res, file) => {
        res[file] = path.resolve(srcPath, file);
        return res;
    }, {});

.....
 module: {
    rules: [
                {
                    test: /\.jsx?$/,
                    exclude: /node_modules/,
                    loader: "babel-loader"
                },
                {
                    test: /\.s?css$/,
                    use: [
                        ENV === "production" ? MiniCssExtractPlugin.loader : {loader: "style-loader"},
                        {
                            loader: "css-loader",
                            options: {
                                url: false,
                                importLoaders: 2
                            }
                        },
                        {
                            loader: "postcss-loader",
                            options: {
                                plugins: function () {
                                    return [autoprefixer("last 2 versions")];
                                }
                            }
                        },
                        {
                            loader: "sass-loader", options: {
                                includePaths: [
                                    srcPath
                                ]
                            }
                        }
                    ]
                },
            ],
},
resolve: {
            extensions: [".js", ".jsx"],
            alias: aliases
        },
```

**Comment by oliverhr** [comment]
Same here, also I try downgrading node-sass to ~4.8.3 but I get the same result so we move to dart-sass/fibers or we downgrade to ~6.0.7 and node sass ^4.9.3.

**Comment by evilebottnawi** [comment]
@oliverhr problem is not in `node-sass`

**Comment by stevenvachon** [comment]
I'm running in this issue with 7.1.0

```scss
@import 'non-existent';
```
throws:
```
internal/streams/legacy.js:57
      throw er; // Unhandled stream error in pipe.
      ^
Error: /non-existent.scss
Module build failed (from ../node_modules/mini-css-extract-plugin/dist/loader.js):
ModuleBuildError: Module build failed (from ../node_modules/sass-loader/lib/loader.js):

'use strict';
^
      Invalid CSS after "'": expected 1 selector or at-rule, was "'use strict';"
```

This is not clear. Something like "non-existent.scss does not exist" would avoid a lot of debugging.

**Comment by hal869** [comment]
Could someone clarify what's not happening that needs to happen to get this fixed?

Trying to follow all the references to find a starting point is a rabbit-hole.

Is it @evilebottnawi needing a response from @jhnns ?

**Comment by evilebottnawi** [comment]
@hal869 we need `webpack` api for set `resolve.extensions` and `resolve.mainFields` and set values inside loader.

**Comment by hal869** [comment]
@evilebottnawi Is there an issue open for that?

**Comment by evilebottnawi** [comment]
@hal869 no issue only private dialog in `slack`, but feel free to create issue in `webpack`

**Comment by aprilmintacpineda** [comment]
I'm experiencing the same problem said by @antonellil. I see this issue hasn't been fixed? My case is this:

folder structure:

```
/ root
| -- webpack.prod.js
| -- src
|     | --  web
|     |      | -- components
|     |      |      | -- styles.scss <-- this is where it's being imported
|     | -- shared
|     |     | -- styles <-- this is where all the imports should resolve to
|     |     |     | -- _colors.scss
```

import syntax: `@import 'colors';`

*_colors.scss*

```scss
$blue: #660099;
```

*config*

```jsx
export default [
  {
     module: [
       // ..
       {
          test: /\.scss$/,
          exclude: /node_modules/,
          use: [
            MiniCssExtractPlugin.loader,
            'css-loader',
            {
              loader: 'sass-loader',
              options: {
                includePaths: [path.join(__dirname, 'src/shared/styles')]
              }
            },
            {
              loader: 'postcss-loader',
              options: {
                plugins: [postcssPresetEnv()]
              }
            }
          ]
        }
     ]
  }
];
```

ERROR:

![image](https://user-images.githubusercontent.com/21032419/47629120-41b6fa00-db74-11e8-9fb9-0b327f69b33d.png)

**Comment by ramonbakker** [comment]
I was also experiencing this issue when I upgraded Angular from version 6.1 to 7.0. I am using Foundation 6.5 and the error went away when I changed `@import 'util';` to `@import './util';` in _settings.scss.

**Comment by evilebottnawi** [comment]
Solving this issue require new api for loaders in `webpack`, it is already in `TODO` list with high priority

**Comment by ahem** [comment]
I am experiencing the same problem as @aprilmintacpineda and it seems to originate from [`webpack/loader-utils`](https://github.com/webpack/loader-utils/blob/master/lib/urlToRequest.js). With a configuration like [described above](https://github.com/webpack-contrib/sass-loader/issues/556#issuecomment-433783527) sass-loader's `importsToResolve` function maps an import like `@import 'colors'` into this list of candidates:
```
[ './_colors.scss',
  './_colors.sass',
  './_colors.css',
  './colors.scss',
  './colors.sass',
  './colors.css',
  'units' ]
```
This shows all the sass candidates are now relative paths, meaning that webpack will fail to find anything. Webpack be able to resolve anything with the last item (the original request) but that will not consider filenames like `_colors.scss` to match.

The reason that the paths have all become relative, even though they weren't in the original import, is that this is just how `loader-utils` work. A possible fix is to prepend the import path with "~" to make loader-utils treat the import as a module import. So this works:

```
@import '~colors'
```

This looks odd and weird to me though, and I don't think I have seen any codebase where people prefix all their module imports with "~".

A possible hackish fix in `sass-loader` might be to prepend the path with `"~"` before calling `urlToRequest` in [importsToResolve](https://github.com/webpack-contrib/sass-loader/blob/master/lib/importsToResolve.js#L18) if it doesn't begin with a `.` or a path separator (`path.sep`).

**Comment by evilebottnawi** [comment]
@ahem can you create minimum reproducible test repo?

**Comment by aprilmintacpineda** [comment]
I have a repository that uses `sass-loader@7.1.0` and works without any of the issues aforementioned in this thread.

https://github.com/aprilmintacpineda/webpack-multi-build-template

**Comment by bencergazda** [comment]
> @evilebottnawi any chance that the fix above[1] will be merged into the main package?
> 
> [bencergazda@3787121](https://github.com/bencergazda/sass-loader/commit/3787121b45ff473bece63ba66f5b3a381d0ebd8e)

@tarjei Sorry, I think I should notice that the commit you have linked in my repo is not my work. I just forked `sass-loader` in order to use some of the other PRs, not sure why GitHub linked that here with my profile (now I have removed my forked repo not to interfere with webpack-contrib/sass-loader). All credits to @jhnns (original commit: https://github.com/webpack-contrib/sass-loader/pull/557/commits/03e5f3ecb34403c5e873a5bb138c9a9ba40a9026). Btw, as I see it has been already merged in April 2018.

**Comment by bencergazda** [comment]
@jhnns @evilebottnawi If this adds new information to the issue, it looks like the issue occurs when the sass `@import` contains an absolute path. Minimal reproducible repo in https://github.com/bencergazda/webpack-sass-broken-imports

**Comment by evilebottnawi** [comment]
It is happens for all `resolve` call (i.e aliasing, resolving file in `node_modules` and etc), we can't override `resolver.extensions` inside `loader` so `js` extension prefer over `scss/sass`

**Comment by bencergazda** [comment]
@evilebottnawi I see, thanks for clarifying this. It is strange that it also fails in case I replace the aliased import path with an absolute path (like `/Users/user/webpack-sass-broken-imports/src/components/component`) Does webpack do resolving in this case?

**Comment by bencergazda** [comment]
@evilebottnawi @jhnns Importing as absolute path in my minimal reproducible repo fails in Mac (`@import "/Users/user/webpack-sass-broken-imports/src/components/component";`), but does work Windows (`@import "C:\webpack-sass-broken-imports\src\components\component";`)

**Comment by evilebottnawi** [comment]
@bencergazda `node-sass` supports windows path in `@import`?

**Comment by michaelgmiller** [comment]
> Solving this issue require new api for loaders in `webpack`, it is already in `TODO` list with high priority

@evilebottnawi do you have a link to the issue tracking this?

**Comment by evilebottnawi** [comment]
@micahlmartin it is internal request (in `slack`), no issue

**Comment by michaelgmiller** [comment]
> @micahlmartin it is internal request (in `slack`), no issue

Filed https://github.com/webpack/webpack/issues/8447 to track this and make sure it gets resolved

**Comment by johncrim** [comment]
This issue is referenced on Stack Overflow:
https://stackoverflow.com/a/53602754/605067

**Comment by evilebottnawi** [comment]
PR https://github.com/webpack/webpack/pull/8452

**Comment by mikesown** [comment]
@evilebottnawi With the webpack fix landed, is it pretty straightforward to fix in sass-loader?

**Comment by evilebottnawi** [comment]
@mikesown need send a PR, i have time on time in near future

**Comment by HairyRabbit** [comment]
any update here ?   :)

**Comment by evilebottnawi** [comment]
@HairyRabbit how it is related to this issue?

**Comment by evilebottnawi** [comment]
And alias works because we send original url to resolver, looks you have invalid url or configuration

**Comment by HairyRabbit** [comment]
yeah, sorry for that, because I found some discussions about the resolves list. 

I try to find some way to let the alias works fine with libsass resolve algorithm like "partials import".

**Comment by evilebottnawi** [comment]
@HairyRabbit feel free to open issue with reproducible test repo

**Comment by evilebottnawi** [comment]
Somebody can create minimum reproducible test repo? I can't reproduce bug in test env :confused: , but i known bug exists.

**Comment by dogoku** [comment]
@evilebottnawi I think @bencergazda shared one above:

https://github.com/bencergazda/webpack-sass-broken-imports

**Comment by evilebottnawi** [comment]
@dogoku this already fixed, but i find reproducible problem, WIP on this

**Comment by nwoltman** [comment]
@evilebottnawi Here's one that's still failing for me:
https://github.com/nwoltman/sass-loader-bug-demo

**Comment by evilebottnawi** [comment]
@nwoltman thanks!

**Comment by evilebottnawi** [comment]
Here two problem:
1. Prefer `js` over `sass`/`scss` extension when you import module (solved here https://github.com/webpack-contrib/sass-loader/pull/645)
2. For better integrating with webpack we use `this.resolve`, but when you setup `includePaths` and use `@import 'globals'` we have a problem with resolving, `node-sass` doesn't provide interface to check file exists in `includePaths` and if you have `node_modules/globals` package, resolver return path this package (and here js file). Feedback welcome :+1:

**Comment by evilebottnawi** [comment]
@nwoltman using `sass` resolution doesn't allow use `alias`, `sass` and `style` fields in `package.json` and etc features

**Comment by evilebottnawi** [comment]
We can require always use `~` from importing from `node_modules`, so `@import globals` doesn't look in `node_modules` and should solve issue, but it is breaking change

**Comment by evilebottnawi** [comment]
Or we can rejeact all resolved files what don't have `sass`/`scss`/`css` extension (some limitation), but it is solve issue for all

**Comment by mikesown** [comment]
@evilebottnawi Thanks for fixing this!!! Now that the first fix is landed, is it possible to push a new version to NPM? I’m excited to

**Comment by evilebottnawi** [comment]
@mikesown no, use `master` now, we need more features/fixes before publish new version

**Comment by evilebottnawi** [comment]
Released planned on next week

**Comment by michaelgmiller** [comment]
> Release planned on next week

Super excited for the release!!

**Comment by vertic4l** [comment]
how many hours left? :)

**Comment by evilebottnawi** [comment]
@vertic4l sorry, a lot of work, also we find the bug in `webpack` in resolve logic, we need fix it before release new version

**Comment by evilebottnawi** [comment]
Works in progress don't worry, it was released in near future

**Comment by evilebottnawi** [comment]
@micahlmartin internal discussion, no issue in bug tracker

**Comment by evilebottnawi** [comment]
Will be released after holidays, sorry for delay, many work, less rest

**Comment by jseegers** [comment]
Looking forward to this release, very happy to see!

**Comment by ovlb** [comment]
Good to know. And take care of yourself! :)

**Comment by mikesown** [comment]
Have the underlying issues in webpack been resolved, or are those still blockers?

**Comment by aendrew** [comment]
I finally was able to fix this issue in my project by adding a `resolve` stanza to my `.scss` rule. 

Example:

```
      {
        test: /.scss$/,
        resolve: {
          extensions: ['.scss', '.sass'],
        },
        use: [
          MiniCssExtractPlugin.loader,
          { loader: 'css-loader', options: { sourceMap: true } },
          { loader: 'postcss-loader', options: { sourceMap: true } },
          {
            loader: 'sass-loader',
            options: {
              sourceMap: true,
              includePaths: ['bower_components'],
            },
          },
        ],
      },
```

Key bit is:

```
        resolve: {
          extensions: ['.scss', '.sass'],
        },
```

**Comment by aendrew** [comment]
I finally was able to fix this issue in my project by adding a `resolve` stanza to my `.scss` rule. 

Example:

```
      {
        test: /.scss$/,
        resolve: {
          extensions: ['.scss', '.sass'],
        },
        use: [
          MiniCssExtractPlugin.loader,
          { loader: 'css-loader', options: { sourceMap: true } },
          { loader: 'postcss-loader', options: { sourceMap: true } },
          {
            loader: 'sass-loader',
            options: {
              sourceMap: true,
              includePaths: ['bower_components'],
            },
          },
        ],
      },
```

Key bit is:

```
        resolve: {
          extensions: ['.scss', '.sass'],
        },
```

**Comment by aprilmintacpineda** [comment]
> I finally was able to fix this issue in my project by adding a `resolve` stanza to my `.scss` rule.
> 
> Example:
> 
> ```
>       {
>         test: /.scss$/,
>         resolve: {
>           extensions: ['.scss', '.sass'],
>         },
>         use: [
>           MiniCssExtractPlugin.loader,
>           { loader: 'css-loader', options: { sourceMap: true } },
>           { loader: 'postcss-loader', options: { sourceMap: true } },
>           {
>             loader: 'sass-loader',
>             options: {
>               sourceMap: true,
>               includePaths: ['bower_components'],
>             },
>           },
>         ],
>       },
> ```
> Key bit is:
> 
> ```
>         resolve: {
>           extensions: ['.scss', '.sass'],
>         },
> ```

Thanks! it worked for me. I was able to upgrade from 6.0.7 to 7.1.0. Didn't know that option exists.

**Comment by evilebottnawi** [comment]
@aendrew you can do this on `sass-loader` position:
```
{
            resolve: {
               extensions: ['.scss', '.sass'],
            },
            loader: 'sass-loader',
            options: {
              sourceMap: true,
              includePaths: ['bower_components'],
            },
          },
```

**Comment by aprilmintacpineda** [comment]
> I finally was able to fix this issue in my project by adding a `resolve` stanza to my `.scss` rule.
> 
> Example:
> 
> ```js
>       {
>         test: /.scss$/,
>         resolve: {
>           extensions: ['.scss', '.sass'],
>         },
>         use: [
>           MiniCssExtractPlugin.loader,
>           { loader: 'css-loader', options: { sourceMap: true } },
>           { loader: 'postcss-loader', options: { sourceMap: true } },
>           {
>             loader: 'sass-loader',
>             options: {
>               sourceMap: true,
>               includePaths: ['bower_components'],
>             },
>           },
>         ],
>       },
> ```
> Key bit is:
> 
> ```js
>         resolve: {
>           extensions: ['.scss', '.sass'],
>         },
> ```

I tried this. It did't work. Same error.



> @aendrew you can do this on `sass-loader` position:
> 
> ```js
> {
>             resolve: {
>                extensions: ['.scss', '.sass'],
>             },
>             loader: 'sass-loader',
>             options: {
>               sourceMap: true,
>               includePaths: ['bower_components'],
>             },
>           },
> ```

I tried this. It didn't work with the following error:

```
✖ ｢wds｣: Invalid configuration object. Webpack has been initialised using a configuration object that does not match the API schema.
 - configuration.module.rules[4].use should be one of these:
   non-empty string | function | object { ident?, loader?, options?, query? } | function | [non-empty string | function | object { ident?, loader?, options?, query? }]
   -> Modifiers applied to the module when rule is matched
   Details:
    * configuration.module.rules[4].use should be a string.
    * configuration.module.rules[4].use should be an instance of function
    * configuration.module.rules[4].use should be an object.
    * configuration.module.rules[4].use should be an instance of function
    * configuration.module.rules[4].use[2] should be a string.
    * configuration.module.rules[4].use[2] should be an instance of function
    * configuration.module.rules[4].use[2] has an unknown property 'resolve'. These properties are valid:
      object { ident?, loader?, options?, query? }
```

I tried:

```js
{
  loader: 'sass-loader',
  options: {
    resolve: {
       extensions: ['.scss', '.sass'],
    },
    sourceMap: true,
    includePaths: ['bower_components'],
  },
},
```

It solved the error above, but resulted to the same error.

```
client:168 ./src/web/components/RightSidebar/loginForm.scss (./node_modules/css-loader/dist/cjs.js!./node_modules/sass-loader/lib/loader.js??ref--8-2!./node_modules/postcss-loader/src??ref--8-3!(webpack)-loader-clean-pragma??ref--8-4!./src/web/components/RightSidebar/loginForm.scss)
Module build failed (from ./node_modules/sass-loader/lib/loader.js):

var colors = require('./colors');
^
      Invalid CSS after "v": expected 1 selector or at-rule, was "var colors = requir"
      in ~/projects/project/node_modules/colors/lib/index.js (line 1, column 1)
```

I think, the key is this like `in ~/projects/project/node_modules/colors/lib/index.js (line 1, column 1)` It seems to have fetched the wrong file.

Given the following config:

```js
{
  test: /\.scss$/,
  exclude: /node_modules/,
  use: [
    isProduction ? MiniCssExtractPlugin.loader : 'style-loader',
    'css-loader',
    {
      loader: 'sass-loader',
      resolve: {
        extensions: ['.scss']
      },
      options: {
        includePaths: [path.join(__dirname, '../src/_shared/styles')],
        functions: {
          staticUrl (_path, done) {
            ...
          },
          url (_path, done) {
            ...
          }
        }
      }
    },
    {
      loader: 'postcss-loader',
      options: {
        plugins: [postcssPresetEnv()]
      }
    },
    {
      loader: 'webpack-loader-clean-pragma',
      options: {
        pragmas
      }
    }
  ]
}
```

I was doing:

```scss
@import 'colors';
```

Which, I was expecting to be fetched from `path.join(__dirname, '../src/_shared/styles')` defined in `includePaths`, not from node_modules.

**Comment by evilebottnawi** [comment]
@larpo hm, should be work, i will investigate, thanks for information

**Comment by aendrew** [comment]
@aprilmintacpineda Have you updated Webpack lately? This leverages a new core Webpack feature that dropped in December. Make sure to update Webpack if you haven't!!!

**Comment by aprilmintacpineda** [comment]
@aendrew all my dependencies are up to date. That's why I want to fix this too, because as of the moment sass-loader is the only one that's out of date.

**Comment by rajuahmmed** [comment]
Is it updated to npm?

**Comment by rajuahmmed** [comment]
@aprilmintacpineda I was facing same issue. I was importing `_colors.scss` with `@import 'colors';` 

Like this
```javascript
{
    loader: 'sass-loader',
    query: {
        includePaths: [ path.join(__dirname, './src/scss') ],
        data: '@import "colors"; @import "variables"; @import "mixins";',
    },
}
```

But when I changed `_colors.scss` to `_color.scss` and
```javascript
{
    loader: 'sass-loader',
    query: {
        includePaths: [ path.join(__dirname, './src/scss') ],
        data: '@import "color"; @import "variables"; @import "mixins";',
    },
}
```

So main issue was with `includePaths`. `includePaths` not working properly. It was importing colors node module into it.

**Comment by aprilmintacpineda** [comment]
Yes, that's exactly what I have pointed out. It's importing stuff from node_modules instead of from what was defined in `includePaths`. I think sass-loader would have problem with this. They wouldn't know where to look first, since I'm sure some others are actually importing sass files from `node_modules`, the question is, how do they know where to find the file without accidentally importing the wrong one when two files have the same name but are on different directories?

Work around is to change the name of one of them.

**Comment by aprilmintacpineda** [comment]
Weird, upon changing the name name from `colors.scss` to `theme-colors.scss`and changing imports from `@import 'colors'` to `@import 'theme-colors'`. It worked as expected, the weird part now is that it's treating SCSS files as if they are JS files.

```
styles.scss:3 Uncaught (in promise) Error: Module build failed (from ./node_modules/sass-loader/lib/loader.js):

'use strict';
^
      Invalid CSS after "'": expected 1 selector or at-rule, was "'use strict';"
      in /Users/me/projects/my-project/node_modules/base/index.js (line 1, column 1)
    at Object../node_modules/css-loader/dist/cjs.js!./node_modules/sass-loader/lib/loader.js?!./node_modules/webpack-loader-clean-pragma/index.js?!./src/web/app.scss (styles.scss:3)
    at __webpack_require__ (bootstrap:63)
    at Object../src/web/app.scss (app.scss:2)
    at __webpack_require__ (bootstrap:63)
    at Module../src/web/App.js (main.1d242e71546b755710db.js:20331)
    at __webpack_require__ (bootstrap:63)
    at Module../src/web/init.js (main.1d242e71546b755710db.js:20803)
    at __webpack_require__ (bootstrap:63)
    at entry.js:24
```

Upon checking the file, it looks like this:

```js
exports = module.exports = require("../../../../node_modules/css-loader/dist/runtime/api.js")(false);
// Module
exports.push([module.id, "/** @format */\n/** @format */\n/** @format */\n.maintenance-view-wrapper .icomoon {\n  color: #ffffff;\n  font-size: 3.313rem;\n  display: block;\n  text-align: center; }\n\n.maintenance-view-wrapper h1 {\n  color: #dab867;\n  font-family: \"dinpro\";\n  font-size: 1.313rem;\n  text-transform: uppercase;\n  text-align: center;\n  margin-bottom: 0.67em; }\n\n.maintenance-view-wrapper p {\n  max-width: 90%;\n  color: #d3dee3;\n  margin: 0 auto;\n  font-size: 0.875rem; }\n", ""]);
```

Here's my config related to `.scss` files. I tried changing but nothing I did cause any change to the error.

```js
{
  test: /\.scss$/,
  exclude: /node_modules/,
  use: [
    isProduction ? MiniCssExtractPlugin.loader : 'style-loader',
    'css-loader',
    {
      loader: 'sass-loader',
      options: {
        includePaths: [path.join(__dirname, '../src/_shared/styles')],
        functions: {
          staticUrl (_path, done) {
            if (process.env.APP_ENV_STATIC_DOMAIN !== '/') {
              done(
                nodeSass.types.String(
                  'url(\'' +
                    helpers.cleanSlashes(process.env.APP_ENV_STATIC_DOMAIN) +
                    '/' +
                    helpers.cleanSlashes(_path.getValue()) +
                    '\')'
                )
              );
            } else {
              done(
                nodeSass.types.String(
                  'url(\'/' + helpers.cleanSlashes(_path.getValue()) + '\')'
                )
              );
            }
          },
          url (_path, done) {
            if (publicUrl !== '/') {
              done(
                nodeSass.types.String(
                  'url(\'/' + publicUrl + helpers.cleanSlashes(_path.getValue()) + '\')'
                )
              );
            } else {
              done(
                nodeSass.types.String(
                  'url(\'/' + helpers.cleanSlashes(_path.getValue()) + '\')'
                )
              );
            }
          }
        }
      }
    },
    {
      loader: 'postcss-loader',
      options: {
        plugins: [postcssPresetEnv()]
      }
    },
    {
      loader: 'webpack-loader-clean-pragma',
      options: {
        pragmas
      }
    }
  ]
}
```

**Comment by evilebottnawi** [comment]
Please don't spam here, known issue, in todo list after update `copy-webpack-plugin` (should be done today)

**Comment by mjankucic-itk** [comment]
Workaround for Angular7+ (unable to revert sass-loader version) & Foundation6+ combination:
Do not import util/util from foundation in your .scss files but separated packages combination (breakpoint, color, direction...) or trick it with **./**util/util. Also foundation.scss and settings.scss imports util inside them and so this files should be avoided or patched locally until sass-loader fix.

**Comment by sridhar-v9** [comment]
Team, 

 When i tried to install the following npm sass fix

   npm install git://github.com/webpack-contrib/sass-loader.git#fix-resolve-conflict --save-dev

getting the following error 
   npm ERR! error: pathspec 'fix-resolve-conflict' did not match any file(s) known to git.

Please provide any comments/suggestions

Thanks

**Comment by sampathkumarm** [comment]
Hi,

Is there any fix for this yet. Below is my webpack config but i still get issues.

package.json
![image](https://user-images.githubusercontent.com/6489309/56076082-43baf100-5d81-11e9-8295-53ec957e977d.png)

Webpack config:
![image](https://user-images.githubusercontent.com/6489309/56076042-d4dd9800-5d80-11e9-916f-45284e4788f5.png)

Importing motion-ui in my .scss file
![image](https://user-images.githubusercontent.com/6489309/56076060-0ce4db00-5d81-11e9-9d32-d64726e24c33.png)

Error:
![image](https://user-images.githubusercontent.com/6489309/56076054-f50d5700-5d80-11e9-91a4-12819f7201d3.png)

Is there any issue with my setup or am i missing anything.

**Comment by apexskier** [comment]
Same issue here with an internal module we use. We publish a package with the files `_vars.scss` and `vars.ts`. I import in sass with the syntax `@import "vars"` which resolves to `vars.ts` in v7 (since resolution now goes through webpack's resolve?).

I'm able to work around for now by changing to `@import "_vars"` explicitly, which webpack doesn't find.

**Comment by bencergazda** [comment]
@apexskier @sampathkumarm Doesn't @aendrew's and @evilebottnawi's solution work for you? https://github.com/webpack-contrib/sass-loader/issues/556#issuecomment-463314416

**Comment by apexskier** [comment]
@bencergazda Yes, it does, and that's what I've ended up using actually (after I commented).

**Comment by sampathkumarm** [comment]
Sorry for the delayed response, I am going to test it and comment back.

**Comment by rj33** [comment]
@evilebottnawi 's fix of adding:
 resolve: {
       extensions: ['.scss', '.sass'],
    },
Did not work for me. As long as I had node_modules in my toplevel webpack  resolve list, it continued to resolve @import "base";  to
node_modules/base/index.js
Changing it to @import "_base";
was the only workaround I could find for this (basically making sure the imported name isn't the same as any node module I have installed under node_modules).

**Comment by richardkeller411** [comment]
Not spamming ( or trying not to ). I get same error with this SCSS...     

@import "theme/variables";
$theme-default: #96cb57;
@import "theme/style";
@import "theme/menu";
@import "theme/responsive";


$theme-default:-dark: darken($theme-default:, 5%) !default;  - Error Here  ( Invalid CSS after "...e-default:-dark": expected 1 selector or at-rule, was ": darken($theme-def")

My Webpack

  webpackConfig.module.rules.push({
      test: /\.s(a|c)ss$/,
      resolve: {
        extensions: [".scss", ".sass"]
      },
      use: [
        "babel-loader",
        "raw-loader",
        "postcss-loader",
        {
          loader: "sass-loader",
          options: {
            includePaths: ["scss", "node_modules"]
              .map(dir => path.join(__dirname, dir))
              .map(gl => glob.sync(gl))
              .reduce((ap, cp) => ap.concat(cp), [])
          }
        }
      ]
    });

**Comment by opr** [comment]
Also experiencing this

webpack config here: https://gist.github.com/opr/adae0b0a062bfa2769be38e13ebeffd0

my files live in `project_root/assets/styles/scss/style.scss` and I also have the following directories: `project_root/assets/styles/scss/helpers` and `project_root/assets/styles/scss/components` - in my style.scss I do `@import helpers/*` however this starts trying to import files from `project_root/helpers` how can I stop this, changing the `test` parameter of my rule doesn't help

**Comment by josephecombs** [comment]
I was experiencing this exact error importing a .css file with sass-loader 7.1.0. 

This is obviously a hacky solution, but **I changed the file extension to .scss** and things started working again.

**Comment by yisding** [comment]
@rj33 I've found adding another line to the resolve block solves the problem of it trying to import base/index.js:

    resolve: {
      extensions: ['.scss', '.sass'],
      mainFiles: []
    },

**Comment by evilebottnawi** [comment]
Can be solved, :+1: https://github.com/webpack/webpack/pull/9422

**Comment by WrapTheCode** [comment]
Solved by:

1. Remove 
`@import 'util/util'; `
**from " _settings.sass"**

2. Move
`@import 'util/util';`
before
`@import 'settings/settings';`
**in "foundation.scss:**

3. And in my sass entrypoint:
```
@import '~foundation-sites/scss/foundation';
@import '~motion-ui/src/motion-ui';
@include foundation-everything;
@include motion-ui-transitions;
@include motion-ui-animations;
```

**Comment by JimBobSquarePants** [comment]
So what's the actual configuration solution?

**Comment by evilebottnawi** [comment]
It is already fixed

**Comment by JimBobSquarePants** [comment]
> It is already fixed

Odd, I saw this very same issue today using 7.3.1

**Comment by evilebottnawi** [comment]
Latest version is `8.0.0` https://github.com/webpack-contrib/sass-loader/blob/master/package.json#L3

**Comment by JimBobSquarePants** [comment]
Thanks, missed that but it seems the issue is still present in `8.0.0`.

![image](https://user-images.githubusercontent.com/385879/64528669-6828f280-d34c-11e9-984f-705acd088130.png)

![image](https://user-images.githubusercontent.com/385879/64528760-9e667200-d34c-11e9-9fde-22cc3ad5ab45.png)

**Comment by evilebottnawi** [comment]
Please create new issue with reproducible test repo

**Comment by dogoku** [comment]
@JimBobSquarePants did u also update your webpack to 4.36.0 or later and your node to 8.9 or later?

**Comment by JimBobSquarePants** [comment]
@dogoku Yeah everything is up to date. 
@evilebottnawi Will do, I'll add a full reproducible sample.

**Comment by sandermarechal** [comment]
I have the same issue on 7.3.1. Is version 7 no longer supported now 8 is out? Lots of projects still depend on 7 (e.g. Symfony Encore Webpack)

**Comment by samayo** [comment]
same issue

**Comment by derWebdesigner** [comment]
@evilebottnawi But only in v8 right? Like @sandermarechal pointed out, the new version of sass-loader is unfortunately not supported yet by many different projects like Symfony Encore Webpack.

**Comment by evilebottnawi** [comment]
@derWebdesigner yes, only new version fix problem, i strongly recommend update `sass-loader`

**Comment by derWebdesigner** [comment]
@evilebottnawi Thank you for your feedback. I would love to update, but then it's not working anymore with Symfony Encore Webpack. I guess I have to open a ticket there that v8 should be supported as soon as it's possible for them.

**Comment by yoonwaiyan** [comment]
Is it possible to fix this for v7 as well as a minor release? As others mentioned there are many packages still depend on v7.
